pkg/controller/certificate-shim/sync.go

```diff
@@ -44,37 +47,46 @@ const (
 	reasonDeleteCertificate = "DeleteCertificate"
 )
 
-var ingressGVK = networkingv1beta1.SchemeGroupVersion.WithKind("Ingress")
+var ingressGVK = networkingv1beta1.SchemeGroupVersion.WithKind("IngressOrGateway")
+var gatewayGVK = gatewayapi.SchemeGroupVersion.WithKind("Gateway")
 
-func (c *controller) Sync(ctx context.Context, ing *networkingv1beta1.Ingress) error {
-	log := logf.WithResource(logf.FromContext(ctx), ing)
+func (c *controller) Sync(ctx context.Context, obj runtime.Object) error {
```

Comment:
This reconciliation function focuses on multiple object types (`Ingress` + `Gateway` + `HTTPRoute` soon).

The good side is that we get a bit of code deduplication.

On the bad side, we are adding making things blurrier to the reader due to having another level of indirection (i.e., the new "generic lister").

I might be missing something: why not have a separate `Sync` function for each kind (`Ingress`, `Gateway`, and soon `HTTPRoute`)? Is it too costly in terms of boilerplate?

Reply:
A couple of thoughts from me: It's written with the future in mind (this could help with supporting multiple ingress versions at once). Also most of the functionality is the same (reading the annotations), the only difference is parsing the reference to TLS secrets

Reply:
For what it's worth, I started this WIP by writing duplicate functions and I thought it was worse.

Reply:
Could we adjust this to be two distinct controllers, but where there is duplicate logic, factor _that_ into standalone library functions instead? This will be easier to maintain long term, including when/if we ever create alternate ingress-shim's too (rather than having to constantly add more complex switching logic into this one controller)

Reply:
I started working on this idea. As a first step, I opened #4149